common/replay/replay_guardian_impl.go

```diff
@@ -0,0 +1,166 @@
+package replay
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/emirpasic/gods/queues/priorityqueue"
+)
+
+var _ ReplayGuardian = &replayGuardian{}
+
+// replayGuardian is an implementation of ReplayGuardian.
+type replayGuardian struct {
+	// The time source. In production use cases, this is likely to just be time.Now.
+	timeSource func() time.Time
+
+	// The maximum amount of time that a request's timestamp can be ahead of the local wall clock time.
+	maxTimeInFuture time.Duration
+
+	// The maximum amount of time that a request's timestamp can be behind the local wall clock time.
+	maxTimeInPast time.Duration
+
+	// A set of hashes that have been observed within the time window.
+	observedHashes map[string]struct{}
+
+	// A queue of observed hashes, ordered by expiration time. Used to prune old hashes.
```

Comment:
> ordered by expiration time

This is technically true, due to how expiration time directly relates to request time. But it would probably be clearer to describe this queue as ordered by the `requestTimestamp`

Reply:
```
	// A queue of observed hashes, ordered by request timestamp. Used to prune old hashes.
```